File: Codility/5_prefix_sums/min_avg_two_slice.cpp
// https://codility.com/programmers/task/min_avg_two_slice/

int solution(vector<int> &A) {
    int minSumEndingHere = A[0] + A[1];
    int countEndingHere = 2;
    double minSoFar = minSumEndingHere / countEndingHere;
    int minStartingIndex = 0;

    for (size_t i = 2; i < A.size(); ++i)
    {
        const double newAvg = static_cast<double>(minSumEndingHere + A[i]) / (countEndingHere + 1);
        const double avgLastTwo = static_cast<double>(A[i] + A[i - 1]) / 2;

        if (avgLastTwo < newAvg)
        {
            minSumEndingHere = A[i] + A[i - 1];
            countEndingHere = 2;
        }
        else
        {
            minSumEndingHere += A[i];
            ++countEndingHere;
        }
        
        const double minAvgEndingHere = static_cast<double>(minSumEndingHere) / countEndingHere;
        if (minAvgEndingHere < minSoFar)
        {
            minSoFar = minAvgEndingHere;
            minStartingIndex = i - countEndingHere + 1;
        }
    }
    
    return minStartingIndex;
}

File: Codility/5_prefix_sums/min_avg_two_slice.py
# https://codility.com/programmers/task/min_avg_two_slice/

def solution(A):
    min_avg_ending_here = float(A[0] + A[1]) / 2
    count_min_ending_here = 2
    min_so_far = min_avg_ending_here
    min_starting_index = 0
    
    for i in xrange(2, len(A)):
        new_avg = (min_avg_ending_here * count_min_ending_here + A[i]) / (count_min_ending_here + 1)
        avg_last_two = float(A[i - 1] + A[i]) / 2
        
        if avg_last_two < new_avg:
            min_avg_ending_here = avg_last_two
            count_min_ending_here = 2
        else:
            min_avg_ending_here = new_avg
            count_min_ending_here += 1
            
        if min_avg_ending_here < min_so_far:
            min_so_far = min_avg_ending_here
            min_starting_index = i - count_min_ending_here + 1
            
    return min_starting_index